### ISSUE TITLE
AI: Switch QLoRA fine-tuning pipeline to Unsloth (2x speed, lower VRAM)

### ISSUE THREAD

**Issue by anchapin** [anchapin/portkit#1322]
## Summary

Replace the vanilla `trl.SFTTrainer` + `bitsandbytes` QLoRA setup in the ml-intern fine-tuning pipeline with **[Unsloth](https://github.com/unslothai/unsloth)**, which provides 2–4x faster training and ~40% lower VRAM usage through custom CUDA kernels — making the training run feasible on RTX 4090 consumer hardware without reducing quality.

## Problem

The current ml-intern fine-tuning plan (documented in `docs/ml_intern_finetuning_prompt.md`) uses a standard QLoRA recipe:
```python
from transformers import AutoModelForCausalLM
from peft import LoraConfig, get_peft_model
from bitsandbytes import BitsAndBytesConfig
from trl import SFTTrainer
```

With Qwen2.5-Coder-7B at `max_seq_length=4096` and batch size 2 + gradient accumulation 8, a 3-epoch run on 1200 training pairs is estimated at ~8–12 hours on a single RTX 4090. With Unsloth's fused attention and optimized LoRA kernels, the same run takes ~2–4 hours with ~30% less VRAM overhead.

## What to do

1. **Install Unsloth** in the ml-intern training environment:
   ```bash
   pip install unsloth
   ```
   Note: Unsloth requires `torch >= 2.1.0` and CUDA 12.1+. Use the conda/pip environment from the training setup.

2. **Replace the model loading code** with Unsloth's FastLanguageModel:
   ```python
   from unsloth import FastLanguageModel
   
   model, tokenizer = FastLanguageModel.from_pretrained(
       model_name="Qwen/Qwen2.5-Coder-7B-Instruct",
       max_seq_length=4096,
       dtype=None,  # auto-detect bfloat16
       load_in_4bit=True,
   )
   
   model = FastLanguageModel.get_peft_model(
       model,
       r=64,
       target_modules=["q_proj", "k_proj", "v_proj", "o_proj", "gate_proj", "up_proj", "down_proj"],
       lora_alpha=128,
       lora_dropout=0.1,
       bias="none",
       use_gradient_checkpointing="unsloth",
       random_state=42,
   )
   ```

3. **Keep `trl.SFTTrainer`** — Unsloth is compatible with it. Only the model loading changes.

4. **Update `docs/ml_intern_finetuning_prompt.md`** to use the Unsloth recipe above as the canonical training recipe.

## Benchmarks (expected)

| Setup | Training time (est.) | VRAM usage |
|-------|---------------------|------------|
| Standard QLoRA (bitsandbytes) | 8–12 hrs | ~20–22 GB |
| Unsloth QLoRA | 2–4 hrs | ~14–16 GB |

## References

- Source document: *AI for Game and Mod Coding* — Section 5 (Training and Fine-Tuning), "Unsloth"
- Unsloth GitHub: https://github.com/unslothai/unsloth
- Unsloth Qwen2.5 notebook: https://colab.research.google.com/github/unslothai/unsloth/blob/main/notebooks/Qwen2.5_Coder_Conversational.ipynb
